docs/eBpfStringsAPI.md

```diff
@@ -0,0 +1,91 @@
+
+## Scenarios
+
+### Finding a token in a header
+
+Assuming an appropriate extension existed to interact with an in-flight HTTP request, a customer wants to search through the Accept-Encoding header, because some older application can't handle gzip appropriately. (I'm sure this is a real use case somewhere.) Error handling is elided in this example for brevity.
+
+```C
+char accept_encoding[] = "Accept-Encoding";
+int ae_len = sizeof(accept_encoding);
+char gzip_name[] = "gzip";
+int gzip_len = sizeof(gzip_name) - 1; // exclude terminal null
+char *header_body = NULL;
+int header_length = ebpfHttpLoadHeader(ctx, accept_encoding, ae_len, &header_body);
+
+char *gzip_location = ebpf_strstr(header_body, header_length, gzip_name, gzip_len);
+// if gzip isn't first, copy everything before it back to the header, then copy everything after it back to the header.
+if (gzip_location > header_body)
+{
+    ebpfHttpSetHeader(ctx, accept_encoding, ae_len, header_body, (gzip_location - header_body));
+}
+
+char *after_header = gzip_location + gzip_len;
+int after_size = header_length - ((gzip_location - header_body) + gzip_len);
+while (after_size > 1)
+{
+    // This loop has to check for at least a terminal null, so it requires a length > 1
+    if (*after_header == ',' || *after_header == ' ')
+    {
+        after_size--;
+        after_header++;
+        continue;
+    }
+    else
+    {
+        break;
+    }
+}
+
+if (after_size > 1)
+{
+    ebpfHttpAppendHeader(ctx, accept_encoding, ae_len, after_header, after_size);
+}
+```
+
+## Functions of interest
+
+We should aim to have sufficient coverage of the core C string functions. The functions that are of interest in particular are listed here; their eBPF counterparts will necessarily need to be somewhat different, especially an example like `strstr()`.
+
+```C
+errno_t strncpy_s(char *restrict dest, size_t dest_size, const char *restrict src, size_t src_count);
+errno_t strncat_s(char *restrict dest, size_t dest_size, const char *restrict src, size_t src_count);
+size_t strnlen_s(const char *str, size_t str_size);
+int strncmp(const char *lhs, const char *rhs, size_t count);
+char *strchr(const char *str, int ch);
+char *strstr(const char *str, const char *substr);
+```
+
+This proposal isn't currently considering wide strings, but if there's appetite for them they'd follow the same pattern.
+
+## Proposed names & prototypes
+
+```C
+errno_t bpf_strcpy(char *restrict dest, size_t dest_size, const char *restrict src, size_t src_count);
+errno_t bpf_strcat(char *restrict dest, size_t dest_size, const char *restrict src, size_t src_count);
+size_t bpf_strlen(const char *str, size_t str_size);
+int bpf_strcmp(const char *lhs, size_t lhs_size, const char *rhs, size_t rhs_size, size_t count);
```

Comment:
What about the cross-platform bpf_strncmp that Linux support?
https://ebpf-docs.dylanreimerink.nl/linux/helper-function/bpf_strncmp/

(Interestingly though, the argument order doesn't match the standard C strncmp: https://en.cppreference.com/w/c/string/byte/strncmp)

Reply:
The definition of `bpf_strncmp` is _weird_. Not only does it not match the standard `strncmp`, it's not even consistent with other buffer-handling functions. Their third parameter is just a string, no declared buffer length. My understanding from conversations with the team was that every buffer had to have its length; it seems the Linux BPF runtime has some different types that it supports in addition to the ones we've already stated here.

I think the revised form of that remark is, I don't think I'd want to pursue implementing their version over one that brings each string's length explicitly.

Reply:
@dthaler I've added `bpf_strtol` and `bpf_strtoul`, since both of those fit the pattern of what we're looking for; the upstream `bpf_strncmp` leaves me uneasy.